lasy/utils/mode_decomposition.py

```diff
@@ -0,0 +1,144 @@
+from lasy.profiles.transverse.transverse_profile import TransverseProfile
+from lasy.profiles.transverse.transverse_profile_from_data import (
+    TransverseProfileFromData,
+)
+from lasy.profiles.transverse.hermite_gaussian_profile import (
+    HermiteGaussianTransverseProfile,
+)
+from lasy.utils.exp_data_utils import find_d4sigma
+
+import numpy as np
+import math
+
+
+def hermite_gauss_decomposition(laserProfile, n_x_max=12, n_y_max=12, res=1e-6):
+    """
+    Decomposes a laser profile into a set of hermite-gaussian modes.
+
+    The function takes either an instance of `TransverseProfile` or an
+    instance of `Laser` (that is, either a transverse profile or the
+    full 3D laser profile defined on a grid). In the case that an
+    instance of `Laser` is passed then the intensity of this profile
+    is projected onto an x-y plane for the decomposition.
+
+    Parameters
+    ----------
+    laserProfile : class instance
+        An instance of a class or sub-class of TransverseLaserProfile
+
+    n_x_max, n_y_max : ints
+        The maximum values of `n_x` and `n_y` out to which the expansion
+        will be performed
+
+    res : float
+        The resolution of grid points in x and y that will be used
+        during the decomposition calculation
+
+    Returns
+    -------
+    weights : dict of floats
+        A dictionary of floats corresponding to the weights of each mode
+        in the decomposition. The keys of the dictionary are tuples
+        corresponding to (`n_x`,`n_y`)
+
+        waist : Beam waist for which the decomposition is calculated.
+            It is computed as the waist for which the weight of order 0 is maximum.
```

Comment:
That's probably my bad, I may have messed up with the indentation in my comment.
```suggestion
    waist : Beam waist for which the decomposition is calculated.
        It is computed as the waist for which the weight of order 0 is maximum.
```